### PR TITLE
stat/distuv: add WithoutReplacement for sampling from a set of intege…

### PR DESCRIPTION
**Pull request by btracey**
…rs without replacement

### CODE REVIEW

File: stat/sampleuv/withoutreplacement.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import (
+	"math/rand"
+	"sort"
+)
+
+// WithoutReplacement samples len(idx) integers from [0, max) without replacement.
+// That is, upon return the elements of idx will be unique integers. If source
+// is non-nil it will be used to generate random numbers, otherwise the default
+// source from the math/rand package will be used.
+func WithoutReplacement(idxs []int, n int, source *rand.Rand) {
```
Comment (kortschak) [inline stat/sampleuv/withoutreplacement.go]: We use `src` elsewhere.

**Comment by btracey** [comment]
I wasn't planning on it being a `Sampler` implementation. This is for `[]int`, not `[]float64`. Nevertheless, it seemed like a useful functionality (at least, I need it), and this seemed a good place to put it. `Weighted` is essentially the weighted version of `WithoutReplacement`.

**Comment by btracey** [comment]
Let me rephrase. I think this is a useful function. `sampleuv` and `samplemv` are mostly about getting a group of samples from a probability distribution (possibly by reducing variance). `Weighted` is here, even though it doesn't fit that, partly because there isn't anywhere better. `WithoutReplacement` is in that boat also. We could instead create a package `sampleset` containing `Weighted` and  `WithoutReplacement`, since both of these are trying to sample from a set of things in various ways.

**Comment by kortschak** [comment]
sampleset might be worthwhile. but ask on gonum-dev.

**Comment by kortschak** [comment]
Though please do change the `source` parameter name. I'll re-approve when you do.

**Comment by btracey** [comment]
Changed the parameter name. I also removed the tests without src specified. I was finding them flaky.

File: stat/sampleuv/withoutreplacement.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+// license that can be found in the LICENSE file.
+
+package sampleuv
+
+import (
+	"math/rand"
+	"sort"
+)
+
+// WithoutReplacement samples len(idx) integers from [0, max) without replacement.
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement.go]: Is `max` `n`?
Also, `idx` is `idxs`. And on the next line.

File: stat/sampleuv/withoutreplacement_test.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+				v := idxs[i]
+				for j := i + 1; j < len(idxs); j++ {
+					if v == idxs[j] {
+						allDiff = false
+						break
+					}
+				}
+			}
+			if !allDiff {
+				_ = cas
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement_test.go]: Remove this?

**Comment by btracey** [comment]
PTAL @vladimir-ch

File: stat/sampleuv/withoutreplacement.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+// license that can be found in the LICENSE file.
+
+package sampleuv
+
+import (
+	"math/rand"
+	"sort"
+)
+
+// WithoutReplacement samples len(idxs) integers from [0, max) without replacement.
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement.go]: I think that you mean `n` where you say `max` (also in the function body comment).

File: stat/sampleuv/withoutreplacement_test.go [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+			}
+		}
+		div := 1 / (float64(test.Trials) * float64(test.K))
+		floats.Scale(div, dist)
+		want := make([]float64, test.N)
+		for i := range want {
+			want[i] = 1 / float64(test.N)
+		}
+		if !floats.EqualApprox(want, dist, test.Tol) {
+			t.Errorf("Cas %d: biased sampling. Want = %v, got = %v", want, dist)
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement_test.go]: Missing `cas` argument.

File: stat/sampleuv/withoutreplacement.go (hunk 4 above)
Comment (btracey) [inline stat/sampleuv/withoutreplacement.go]: Oh my gosh, sorry!

File: stat/sampleuv/withoutreplacement_test.go (hunk 5 above)
Comment (btracey) [inline stat/sampleuv/withoutreplacement_test.go]: Done.

File: stat/sampleuv/withoutreplacement.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+// source from the math/rand package will be used.
+func WithoutReplacement(idxs []int, n int, src *rand.Rand) {
+	if len(idxs) == 0 {
+		panic("withoutreplacement: zero length input")
+	}
+
+	// There are two algorithms. One is to generate a random permutation
+	// and take the first len(idxs) elements. The second is to generate
+	// individual random numbers for each element and check uniqueness. The first
+	// method scales as O(max), and the second scales as O(len(idxs)^2). Choose
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement.go]: I'm sorry, `max -> n` also here :-)

File: stat/sampleuv/withoutreplacement.go [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+package sampleuv
+
+import (
+	"math/rand"
+	"sort"
+)
+
+// WithoutReplacement samples len(idxs) integers from [0, n) without replacement.
+// That is, upon return the elements of idxs will be unique integers. If source
```
Comment (vladimir-ch) [inline stat/sampleuv/withoutreplacement.go]: I'm afraid to ask in case I'm missing something obvious: if `n < len(idxs)`, can the elements of idxs be unique?

Comment (btracey) [inline stat/sampleuv/withoutreplacement.go]: Added panic.

File: stat/sampleuv/withoutreplacement.go (hunk 6 above)
Comment (btracey) [inline stat/sampleuv/withoutreplacement.go]: Done.
